modules/text2vec-openai/vectorizer/objects.go

```diff
@@ -87,3 +115,181 @@ func (v *Vectorizer) getVectorizationConfig(cfg moduletools.ClassConfig) ent.Vec
+//
+// On the high level it has the following steps:
+//  1. It receives a batch job
+//  2. It splits the job into smaller vectorizer-batches if the token limit is reached. Note that objects from different
+//     batches are not mixed with each other to simplify returning the vectors.
+//  3. It sends the smaller batches to the vectorizer
+func (v *Vectorizer) batchWorker() {
+	rateLimit := &ent.RateLimits{}
+	texts := make([]string, 0, 100)
+	firstRequest := true
+	timePerToken := 0.0
+	batchTookInS := float64(0)
+BatchLoop:
+	for job := range v.jobQueueCh {
+		// the total batch should not take longer than 60s to avoid timeouts. We will only use 40s here to be safe
+		batchStart := time.Now()
+
+		objCounter := 0
+		vecBatchOffset := 0 // index of first object in current vectorizer-batch
+		tokensInCurrentBatch := 0
+		texts = texts[:0]
+
+		conf := v.getVectorizationConfig(job.cfg)
+
+		// we don't know the current rate limits without a request => send a small one
+		for firstRequest {
+			var err error
+			rateLimit, err = v.makeRequest(job, job.texts[:1], conf, vecBatchOffset)
+			if err != nil {
+				for j := 0; j < len(job.texts); j++ {
+					job.errs[j] = err
+				}
+				job.wg.Done()
+				continue BatchLoop
+			}
+			objCounter++
+			vecBatchOffset++
+			firstRequest = false
+		}
+
+		for objCounter < len(job.texts) {
+			if job.skipObject[objCounter] {
+				objCounter++
+				continue
+			}
+
+			if job.tokens[objCounter] > rateLimit.LimitTokens {
+				job.errs[objCounter] = fmt.Errorf("text too long for vectorization")
+				objCounter++
+				continue
+			}
+
+			// add objects to the current vectorizer-batch until the remaining tokens are used up or other limits are reached
+			text := job.texts[objCounter]
+			if float32(tokensInCurrentBatch+job.tokens[objCounter]) < 0.95*float32(rateLimit.RemainingTokens) && (timePerToken*float64(tokensInCurrentBatch) < MaxBatchTime.Seconds()/4) && len(texts) < MaxObjectsPerBatch {
+				tokensInCurrentBatch += job.tokens[objCounter]
+				texts = append(texts, text)
+				objCounter++
+				if objCounter < len(job.texts) {
+					continue
+				}
+			}
+
+			// if a single object is larger than the current token limit we need to wait until the token limit refreshes
+			// enough to be able to handle the object. This assumes that the tokenLimit refreshes linearly which is true
+			// for openAI, but needs to be checked for other providers
+			if len(texts) == 0 {
+				fractionOfTotalLimit := float32(rateLimit.LimitTokens) / float32(job.tokens[objCounter])
+				sleepTime := time.Duration(float32(rateLimit.ResetTokens)*fractionOfTotalLimit+1) * time.Second
+				if time.Since(batchStart)+sleepTime > MaxBatchTime {
+					time.Sleep(sleepTime)
+					rateLimit.RemainingTokens += int(float32(rateLimit.LimitTokens) * fractionOfTotalLimit)
+				} else {
+					job.errs[objCounter] = fmt.Errorf("text too long for vectorization. Cannot wait for token refresh due to time limit")
+					objCounter++
+				}
```

Comment:
> shouldn't sleep be called only if it does not exceed MaxBatchTime (so < instead >)?

yep, fixed - thanks!

> if MaxBatchTime is exceeded, shouldn't all remaining objects in batch fail with error and batch be finished?

So there are a few things: We do not know the total time we have for a batch. There is a server-timeout (context, so we do not need to worry about), client timeout (we do not know that) and there might be a number of servers between client and server that can have their own timeout settings (often 60s, but we do not know that). So my thinking was to avoid time-expensive operations when we get close to the time-limit but not error all remaining objects. So rather a soft-limit than a hard limit, but I am not sure if this is the best way of handling it 🤷 

I'll add another comment for this